server/src/main/java/org/elasticsearch/ingest/TrackingResultProcessor.java

```diff
@@ -67,19 +66,35 @@ public String getTag() {
         return actualProcessor.getTag();
     }
 
-    public static CompoundProcessor decorate(CompoundProcessor compoundProcessor, List<SimulateProcessorResult> processorResultList) {
+    public static CompoundProcessor decorate(CompoundProcessor compoundProcessor, List<SimulateProcessorResult> processorResultList,
+                                             Set<PipelineProcessor> pipelinesSeen) {
         List<Processor> processors = new ArrayList<>(compoundProcessor.getProcessors().size());
         for (Processor processor : compoundProcessor.getProcessors()) {
-            if (processor instanceof CompoundProcessor) {
-                processors.add(decorate((CompoundProcessor) processor, processorResultList));
+            if (processor instanceof PipelineProcessor) {
+                PipelineProcessor pipelineProcessor = ((PipelineProcessor) processor);
+                if (pipelinesSeen.add(pipelineProcessor) == false) {
+                    throw new IllegalStateException("Recursive invocation of pipeline ["
```

Comment:
I think this is not correct because it would also throw for repeated invocation of the same pipeline right? (Since `PipelineProcessor` doesn't implement `equals` you will probably only see this in some crazier scenarios (A calls B calls C and then another step of just B which would call C again making it throw on seeing C twice even though it's not a recursive invocation)) See #33419 for more here. I think the easiest fix to get out of that problem is to do the same thing that the pipeline processor does and simply track the current stack of pipelines in `pipelinesSeen` instead of all the pipelines ever seen (in this case this means just removing a pipeline proc from the `pipelinesSeen` after it has been unwrapped)

Reply:
good catch.  fixed and updated in latest commit.

Reply:
thanks looks good :)